### PR TITLE
Fail hard when `--concurrency` is set to invalid values

### LINKED ISSUE #1476: Fail when `--concurrency` is set to `0`
Inspired by #1463. `--concurrency=0` currently means AVA spins up as many workers as there are test files, and when #1467 lands it means AVA spins up as many workers as there are CPU cores. In either case, we should refuse to run tests when the value is set to `0` since AVA's behavior won't match user expectations.

See a similar test in `lib/cli.js`: https://github.com/avajs/ava/blob/a0d5b373a19e9eb0022ad2936b46a1608f11b53f/lib/cli.js#L121

### PR DESCRIPTION
**Pull request by tdeschryver**
Fixes #1476.

**Changes:**
- Added `0` check.
- Changed the message if  `concurrency` was set to an empty string, because the error message seemed a bit weird.
- Introduced an extra check to validate if `concurrency` is an integer.
- Changed default `concurrency` from `0` to `1` if no `concurrency` is given.

### CODE REVIEW

File: lib/cli.js [hunk 1]
```diff
@@ -119,7 +119,15 @@ exports.run = () => {
 	}
 
 	if (cli.flags.concurrency === '') {
-		throw new Error(colors.error(figures.cross) + ' The --concurrency and -c flags must be provided the maximum number of test files to run at once.');
+		throw new Error(colors.error(figures.cross) + ' The --concurrency and -c flags must be provided.');
+	}
+
+	if (cli.flags.concurrency && isNaN(parseInt(cli.flags.concurrency, 10))) {
```
Comment (novemberborn) [inline lib/cli.js]: I think this should be `Number.isInteger(Number.parseFloat('cli.flags.concurrency'))`. That way even values like `9.5` are rejected, which parse to `9` with `parseInt()`.

File: lib/cli.js [hunk 2]
```diff
@@ -142,7 +150,7 @@ exports.run = () => {
 		resolveTestsFrom: cli.input.length === 0 ? projectDir : process.cwd(),
 		projectDir,
 		timeout: conf.timeout,
-		concurrency: conf.concurrency ? parseInt(conf.concurrency, 10) : 0,
+		concurrency: conf.concurrency ? parseInt(conf.concurrency, 10) : 1,
```
Comment (novemberborn) [inline lib/cli.js]: The default should remain `0`. This allows AVA to pick an optimal value for your machine, see https://github.com/avajs/ava/blob/465fcecc9ae0d3274d4d41d3baaca241d6a40130/api.js#L164:L167.

Comment (tdeschryver) [inline lib/cli.js]: Oh sorry, I assumed the default had to change, because you wouldn't be able to set it to `0`.

File: lib/cli.js (hunk 1 above)
Comment (tdeschryver) [inline lib/cli.js]: Yep! Didn't thought about this case.

**Comment by tdeschryver** [comment]
@novemberborn seems fair to me. I'll remove the `0` check.

**Comment by tdeschryver** [comment]
@novemberborn would a check on a negative input be useful here? :thinking:

File: lib/cli.js [hunk 3]
```diff
@@ -119,7 +119,11 @@ exports.run = () => {
 	}
 
 	if (cli.flags.concurrency === '') {
-		throw new Error(colors.error(figures.cross) + ' The --concurrency and -c flags must be provided the maximum number of test files to run at once.');
+		throw new Error(colors.error(figures.cross) + ' The --concurrency and -c flags must be provided.');
+	}
+
+	if (cli.flags.concurrency && !Number.isInteger(Number.parseFloat(cli.flags.concurrency, 10))) {
```
Comment (novemberborn) [inline lib/cli.js]: `parseFloat` doesn't take a radix.

File: lib/cli.js [hunk 4]
```diff
@@ -119,7 +119,12 @@ exports.run = () => {
 	}
 
 	if (cli.flags.concurrency === '') {
-		throw new Error(colors.error(figures.cross) + ' The --concurrency and -c flags must be provided the maximum number of test files to run at once.');
+		throw new Error(colors.error(figures.cross) + ' The --concurrency and -c flags must be provided.');
+	}
+
+	if (cli.flags.concurrency &&
+		(!Number.isInteger(Number.parseFloat(cli.flags.concurrency)) || parseInt(cli.flags.concurrency, 10) < 0)) {
+		throw new Error(colors.error(figures.cross) + ' The --concurrency and -c flags must be a positive integer.');
```
Comment (novemberborn) [inline lib/cli.js]: This should be `nonnegative integer`, since `0` is allowed. I feel a bit bad about this nitpick ☹️

Comment (tdeschryver) [inline lib/cli.js]: No problem! I should be the one saying sorry...

**Comment by novemberborn** [comment]
Thank you @tdeschryver!

**Pull request merged**